File: shared/src/timeline-icons.ts
/**
 * Shared icon-key system for timeline entries.
 *
 * Each platform maps the abstract key to its native form:
 *   - Apple: SF Symbols (`checkmark.circle.fill`, etc.)
 *   - Android tablet: Material Icons (`Icons.Filled.CheckCircle`)
 *   - Android e-ink: ASCII bracket markers (`[OK]`, `[T ]`) — high black coverage,
 *     no thin glyph strokes that ghost on partial refresh.
 *
 * This is the single source of truth. If you add a new key, add the platform
 * mapping in the same commit (apple TimelineStripView.swift,
 * android TimelineStrip.kt + EinkTimelinePanel.kt).
 */

import type { TimelineEntry, TimelineEntryType } from './timeline.js';

export type TimelineIconKey =
  | 'success'   // chat completed, tool resolved
  | 'error'     // any error
  | 'running'   // active turn / chat in progress
  | 'awaiting'  // waiting on user (pending tool, chat_start without completion)
  | 'tool'      // tool execution
  | 'model'     // model call/response
  | 'user'      // user action / prompt
  | 'task'      // task hierarchy header
  | 'scheduled' // scheduled work
  | 'memory';   // memory recall

/**
 * Map a timeline entry to its semantic icon key.
 * Looks at both the entry type and (for tool_request) the approval status.
 */
export function timelineIconKey(entry: Pick<TimelineEntry, 'type' | 'status'>): TimelineIconKey {
  switch (entry.type) {
    case 'task_start':
    case 'task_end':
      return 'task';
    case 'task_milestone':
      return 'success';
    case 'chat_start':
      return 'running';
    case 'chat_end':
    case 'chat_response':
    case 'model_response':
      return 'success';
    case 'model_call':
      return 'model';
    case 'tool_request':
      if (entry.status === 'approved') return 'success';
      if (entry.status === 'denied') return 'error';
      // `abandoned` and `pending` share the amber "no decision was made" icon.
      // They are the same fact at two points in time — one still open, one
      // closed without an answer — and neither is a refusal.
      return 'awaiting';
    case 'tool_resolved':
      // The resolution row carries the SAME status as the request it closes, so
      // it must branch on it. Returning 'success' unconditionally (as this did)
      // drew a green check on a denial and on an abandonment alike — the one
      // row whose entire job is to say how the approval ended was the one row
      // that could not say it.
      if (entry.status === 'denied') return 'error';
      if (entry.status === 'abandoned') return 'awaiting';
      return 'success';
    case 'tool_exec':
      return 'tool';
    case 'error':
      return 'error';
    case 'user_action':
      return 'user';
    case 'scheduled':
      return 'scheduled';
    case 'memory_recall':
      return 'memory';
    case 'eval_result':
      return entry.status === 'denied' ? 'error' : 'success';
    default:
      return 'running';
  }
}

/** All known icon keys, in stable order (used for tests/registries). */
export const TIMELINE_ICON_KEYS: readonly TimelineIconKey[] = [
  'success', 'error', 'running', 'awaiting',
  'tool', 'model', 'user', 'task',
  'scheduled', 'memory',
] as const;

/**
 * Staleness cap for the in-flight task spinner: a `task_start` older than
 * this without a matching `task_end` is treated as resolved-but-orphaned
 * (daemon killed mid-task / hook delivery race / session_end early-return).
 * The daemon-side orphan reaper eventually upserts a synthetic task_end, but
 * this UI guard is independent so a stale row reads as completed even before
 * the reaper runs. Value matches Apple's `inFlightTaskMaxAgeSec`.
 */
export const IN_FLIGHT_TASK_MAX_AGE_MS = 10 * 60 * 1000;

/**
 * True when `entry` is a `task_start` whose matching `task_end` (same `taskId`)
 * has not yet appeared among `siblings` and the row is younger than
 * `IN_FLIGHT_TASK_MAX_AGE_MS`. Lets clients render in-flight task hierarchy
 * markers with the rotating "running" treatment instead of the static `task`
 * icon. Mirrored in Apple `timelineIsInFlightTask` and Android
 * `isInFlightTask` — update all three in the same commit.
 */
export function isInFlightTask(
  entry: Pick<TimelineEntry, 'type' | 'taskId'> & Partial<Pick<TimelineEntry, 'ts'>>,
  siblings: ReadonlyArray<Pick<TimelineEntry, 'type' | 'taskId'>>,
  nowMs?: number,
): boolean {
  if (entry.type !== 'task_start') return false;
  if (!entry.taskId) return false;
  for (const s of siblings) {
    if (s.type === 'task_end' && s.taskId === entry.taskId) return false;
  }
  if (entry.ts != null && (nowMs ?? Date.now()) - entry.ts > IN_FLIGHT_TASK_MAX_AGE_MS) {
    return false;
  }
  return true;
}

/**
 * Age cap for the rotating "running" treatment on turn rows. Turn-completion
 * signals are best-effort (Stop hook, transcript tail, PTY marker) and can be
 * lost across daemon handoffs — a chat_start older than this stops spinning
 * even without an explicit completion so dead turns don't animate forever.
 * Matches Apple's `chatStartMaxAgeSec` (TimelineStripView.swift). Task
 * hierarchy markers are handled by the daemon-side orphan reaper plus Apple's
 * own in-flight cap.
 */
export const ROTATING_ENTRY_MAX_AGE_MS = 10 * 60 * 1000;

type RotatingSibling = Pick<TimelineEntry, 'type' | 'taskId'> &
  Partial<Pick<TimelineEntry, 'sessionId' | 'ts'>>;

/** Same-session test mirroring the turn-merge rule: both ids equal, or both
 *  absent (legacy single-session emitters). */
function sameRotatingSession(a?: string, b?: string): boolean {
  if (!a && !b) return true;
  return !!a && !!b && a === b;
}

/**
 * True when a turn row should rotate its leading icon. Combines the static
 * `running` icon-key (chat_start, unknown types) with the in-flight
 * task-hierarchy signal so an open `task_start` also spins until its
 * `task_end` arrives. `siblings` should be the entries the row sees in its
 * group/list — passing `[]` falls back to icon-key only.
 *
 * A chat_start only spins while its turn is plausibly still open:
 *   - no later same-session completion (chat_response / chat_end / model_response),
 *   - no later same-session chat_start (a new prompt supersedes the turn even
 *     when its completion signal was lost — Stop hooks are best-effort),
 *   - younger than ROTATING_ENTRY_MAX_AGE_MS.
 * Mirrored in Apple TimelineStripView.swift and Android TimelineIcons.kt —
 * update all three in the same commit.
 */
export function isRotatingEntry(
  entry: Pick<TimelineEntry, 'type' | 'status' | 'taskId'> &
    Partial<Pick<TimelineEntry, 'sessionId' | 'ts'>>,
  siblings: ReadonlyArray<RotatingSibling>,
  nowMs?: number,
): boolean {
  if (timelineIconKey(entry) === 'running') {
    if (entry.type !== 'chat_start') return true;
    const ts = entry.ts;
    if (ts != null) {
      const now = nowMs ?? Date.now();
      if (now - ts > ROTATING_ENTRY_MAX_AGE_MS) return false;
      for (const s of siblings) {
        if (s.ts == null || s.ts < ts) continue;
        if (!sameRotatingSession(entry.sessionId, s.sessionId)) continue;
        // `error` closes a turn as surely as a response does — a failed Codex
        // request emits one instead of a reply, and leaving it off this list
        // kept the spinner turning next to the error that explained it.
        if (s.type === 'chat_response' || s.type === 'chat_end'
            || s.type === 'model_response' || s.type === 'error') return false;
        if (s.type === 'chat_start' && s.ts > ts) return false;
      }
    }
    return true;
  }
  return isInFlightTask(entry, siblings, nowMs);
}

/**
 * E-ink ASCII glyphs — bracket-padded so total width is constant (4 chars)
 * for column alignment on monospace bitmap fonts.
 */
export const EINK_ICON_GLYPHS: Record<TimelineIconKey, string> = {
  success:   '[OK]',
  error:     '[!!]',
  running:   '[..]',
  awaiting:  '[??]',
  tool:      '[T ]',
  model:     '[M ]',
  user:      '[U ]',
  task:      '[==]',
  scheduled: '[S ]',
  memory:    '[~ ]',
};

/**
 * Whether a timeline entry type carries a content body worth showing in the
 * detail pane. task_start/task_end/task_milestone are hierarchy markers —
 * their `raw` is the summary, no extra detail expected.
 */
export function entryHasDetailBody(type: TimelineEntryType): boolean {
  return type !== 'task_start' && type !== 'task_end' && type !== 'task_milestone';
}

File: android/app/src/main/kotlin/dev/agentdeck/ui/timeline/TimelineIcons.kt
package dev.agentdeck.ui.timeline

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Alarm
import androidx.compose.material.icons.filled.Build
import androidx.compose.material.icons.filled.Checklist
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.HourglassTop
import androidx.compose.material.icons.filled.Memory
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.Psychology
import androidx.compose.material.icons.filled.Sync
import androidx.compose.ui.graphics.vector.ImageVector
import dev.agentdeck.state.TimelineEntry

/**
 * Semantic icon key for a timeline entry. Mirrors `shared/src/timeline-icons.ts`
 * (`TimelineIconKey`) and `apple/AgentDeck/UI/Monitor/TimelineStripView.swift`
 * (`TimelineIconKey`). Single source of truth for the abstract semantics.
 *
 * Each platform maps the key to its native form. On Android tablet we use
 * Material Icons (vector). On Android e-ink we use ASCII bracket markers
 * (see [einkGlyph]) — Material vector glyphs render at thin strokes after
 * 1-bit dither and ghost on partial refresh.
 */
enum class TimelineIconKey {
    Success, Error, Running, Awaiting, Tool, Model, User, Task, Scheduled, Memory;

    /** Material Icon for tablet/full-color surfaces. */
    val materialIcon: ImageVector
        get() = when (this) {
            Success -> Icons.Filled.CheckCircle
            Error -> Icons.Filled.Error
            Running -> Icons.Filled.Sync
            Awaiting -> Icons.Filled.HourglassTop
            Tool -> Icons.Filled.Build
            Model -> Icons.Filled.Psychology
            User -> Icons.Filled.Person
            Task -> Icons.Filled.Checklist
            Scheduled -> Icons.Filled.Alarm
            Memory -> Icons.Filled.Memory
        }

    /** ASCII bracket glyph for e-ink displays. Constant 4-char width for
     *  monospace alignment; high black coverage so the panel doesn't ghost. */
    val einkGlyph: String
        get() = when (this) {
            Success -> "[OK]"
            Error -> "[!!]"
            Running -> "[..]"
            Awaiting -> "[??]"
            Tool -> "[T ]"
            Model -> "[M ]"
            User -> "[U ]"
            Task -> "[==]"
            Scheduled -> "[S ]"
            Memory -> "[~ ]"
        }
}

/**
 * Resolve the icon key for a timeline entry by type + status.
 * Mirrors `timelineIconKey()` in shared/src/timeline-icons.ts.
 */
fun timelineIconKey(type: String, status: String? = null): TimelineIconKey = when (type) {
    "task_start", "task_end" -> TimelineIconKey.Task
    "task_milestone" -> TimelineIconKey.Success
    "chat_start" -> TimelineIconKey.Running
    "chat_end", "chat_response", "model_response" -> TimelineIconKey.Success
    "model_call" -> TimelineIconKey.Model
    // `abandoned` shares the amber no-decision icon with `pending`: they are the
    // same fact at two points in time (still open / closed unanswered) and
    // neither is a refusal.
    "tool_request" -> when (status) {
        "approved" -> TimelineIconKey.Success
        "denied" -> TimelineIconKey.Error
        else -> TimelineIconKey.Awaiting
    }
    // The resolution row carries the SAME status as the request it closes, so it
    // must branch on it. Returning Success unconditionally drew a green check on
    // a denial and on an abandonment alike.
    "tool_resolved" -> when (status) {
        "denied" -> TimelineIconKey.Error
        "abandoned" -> TimelineIconKey.Awaiting
        else -> TimelineIconKey.Success
    }
    "tool_exec" -> TimelineIconKey.Tool
    "error" -> TimelineIconKey.Error
    "user_action" -> TimelineIconKey.User
    "scheduled" -> TimelineIconKey.Scheduled
    "memory_recall" -> TimelineIconKey.Memory
    "eval_result" -> if (status == "denied") TimelineIconKey.Error else TimelineIconKey.Success
    else -> TimelineIconKey.Running
}

/** Whether this entry type carries detail-pane body content (not a hierarchy marker). */
fun entryHasDetailBody(type: String): Boolean =
    type != "task_start" && type != "task_end" && type != "task_milestone"

/**
 * Staleness cap for the in-flight task spinner: a `task_start` older than
 * this without a matching `task_end` is treated as resolved-but-orphaned.
 * Mirrors `IN_FLIGHT_TASK_MAX_AGE_MS` in shared/src/timeline-icons.ts and
 * Apple `inFlightTaskMaxAgeSec` — without it, tablet task rows spun forever
 * on orphaned tasks that macOS already rendered as completed.
 */
const val IN_FLIGHT_TASK_MAX_AGE_MS = 10 * 60 * 1000L

/**
 * True when [entry] is a `task_start` whose matching `task_end` (same
 * `taskId`) hasn't yet appeared in [siblings] and the row is younger than
 * [IN_FLIGHT_TASK_MAX_AGE_MS]. Mirrors `isInFlightTask` in
 * shared/src/timeline-icons.ts — used to spin the leading icon for in-flight
 * task hierarchy markers instead of the static `task` glyph.
 */
fun isInFlightTask(
    entry: TimelineEntry,
    siblings: List<TimelineEntry>,
    nowMs: Long = System.currentTimeMillis(),
): Boolean {
    if (entry.type != "task_start") return false
    val taskId = entry.taskId ?: return false
    if (taskId.isEmpty()) return false
    for (s in siblings) {
        if (s.type == "task_end" && s.taskId == taskId) return false
    }
    if (nowMs - entry.timestamp > IN_FLIGHT_TASK_MAX_AGE_MS) return false
    return true
}

/**
 * Age cap for the rotating "running" treatment on turn rows. Turn-completion
 * signals are best-effort (Stop hook, transcript tail, PTY marker) and can be
 * lost across daemon handoffs — a chat_start older than this stops spinning
 * even without an explicit completion. Mirrors ROTATING_ENTRY_MAX_AGE_MS in
 * shared/src/timeline-icons.ts and `chatStartMaxAgeSec` on Apple.
 */
const val ROTATING_ENTRY_MAX_AGE_MS = 10 * 60 * 1000L

/** Same-session test mirroring the turn-merge rule: both ids equal, or both
 *  absent (legacy single-session emitters). */
private fun sameRotatingSession(a: String?, b: String?): Boolean =
    if (a.isNullOrEmpty() && b.isNullOrEmpty()) true
    else !a.isNullOrEmpty() && !b.isNullOrEmpty() && a == b

/**
 * True when a turn row should rotate its leading icon. Combines the
 * `Running` icon-key (chat_start, unknown types) with the in-flight task
 * hierarchy signal so an open `task_start` also spins until its `task_end`
 * arrives. Mirrors `isRotatingEntry` in shared/src/timeline-icons.ts.
 *
 * A chat_start only spins while its turn is plausibly still open:
 *   - no later same-session completion (chat_response/chat_end/model_response),
 *   - no later same-session chat_start (a new prompt supersedes the turn even
 *     when its completion signal was lost — Stop hooks are best-effort),
 *   - younger than [ROTATING_ENTRY_MAX_AGE_MS].
 */
/**
 * True when a later same-session completion row exists for this chat_start —
 * lets an *unmerged* turn row (its response didn't fold in, e.g. absorbed
 * elsewhere or filtered) still swap to the completed glyph instead of
 * presenting a finished turn as running.
 */
fun turnHasLaterCompletion(entry: TimelineEntry, siblings: List<TimelineEntry>): Boolean {
    if (entry.type != "chat_start") return false
    for (s in siblings) {
        if (s.timestamp < entry.timestamp) continue
        if (!sameRotatingSession(entry.sessionId, s.sessionId)) continue
        // `error` closes a turn as surely as a response does (mirrors Swift/TS).
        if (s.type == "chat_response" || s.type == "chat_end" ||
            s.type == "model_response" || s.type == "error") return true
    }
    return false
}

fun isRotatingEntry(
    entry: TimelineEntry,
    siblings: List<TimelineEntry>,
    nowMs: Long = System.currentTimeMillis(),
): Boolean {
    if (timelineIconKey(entry.type, entry.status) == TimelineIconKey.Running) {
        if (entry.type != "chat_start") return true
        val ts = entry.timestamp
        if (nowMs - ts > ROTATING_ENTRY_MAX_AGE_MS) return false
        for (s in siblings) {
            if (s.timestamp < ts) continue
            if (!sameRotatingSession(entry.sessionId, s.sessionId)) continue
            if (s.type == "chat_response" || s.type == "chat_end" || s.type == "model_response") return false
            if (s.type == "chat_start" && s.timestamp > ts) return false
        }
        return true
    }
    return isInFlightTask(entry, siblings, nowMs)
}
